rfcs/xxxx-collage.md

```diff
@@ -0,0 +1,987 @@
+- `FuseOps` is generally considered too inflexible, and we've sought a more flexible way to express target-dependent
+  fusion rules.
+
+So in the same way `DFPattern`s provide a small library of 'base' and 'combinator' pattern rules supporting a wide
+variety of examples, we seek the same economy and flexibility from `PartitionRule`s. 
+
+An obvious question is whether all partition rules should be expressed with `DFPattern`s, possibly by extending
+the `DFPattern` library itself. Indeed, though it does not appear to be used in prod, the `DominatorPattern` is
+an attempt to use `DFPattern`s to subsume the existing `FuseOps` machinery. We actually went down this path but
+decided to back out:
+- Since we are interested in searching over possibly overlapping candidate partitions we'd need the `DFPattern`
+  machinery to all enumeration over all matching sub-expressions. That would require a rewrite of the
+  `DFPatternMatcher`.
+- Some of the more subtle fusion rules are difficult to express as patterns.
+- `DFPattern`s are widely used outside of just partitioning, so any change would need to ensure no efficiency
+  or cognitive overhead for those common cases.
+
+That pushed us to the present design, which builds on `DFPatterns`, but introduces new 'base' and 'combinator'
+partition rules which can be combined to match the desired partition flavors. The 'base' rules produce candidates
+from the dataflow graph directly (eg all sub-graphs matching a given `DFPattern`). The 'combinator' rules 
+combine the candidates found by one or more sub-rules into a new set of candidates (eg by indicating the 
+sub-candidates should be wrapped by a `"Composite"` function, or touching sub-candidates should be unioned
+together).
+
+The base rules are:
+
+- `DFPatternPartitionRule`: Given a `DFPattern` and expression predicate, produces a candidate for every sub-graph
+  matched by the pattern and predicate. Unlike the `PatternRewriter`, candidates are free to overlap. Mainly used
+  to bring BYOC patterns into the Collage framework.
+- `OpPredicatePartitionRule`: Given an attribute name, produces a candidate for every call to a primitive Relay
+  operator where the operator i) has predicate bound to that attribute which ii) returns true given the
+  call sub-expression. Generally this will result in a singleton sub-graph containing only the call, but it may also
+  pull in constant arguments to the call should they be required. Used to bring BYOC operator predicates into the
+  Collage framework.
+- `OpCallByKindPartitionRule`: Uses the `"TOpPattern"` attribute provided for every Relay operator to produce a
+  candidate for every call to a 'fusable Relay operator'. Used to select the operators which `FuseOps` will consider
+  parts of kernels.
+
+The combinator rules are:
+
+- `CompositePartitionRule`: Indicates all sub-candidates matched by the sub-rule should be wrapped by a `"Composite"`
```

Comment:
How about we use base-candidates and combinator-candidates rather than introducing another term "sub-candidates" ?

Reply:
I'd take those to mean 'candidates produced by a {base,combinator} rule' which isn't the intent of the term. In particular, 'combinator' rules need to be nested, so the 'sub-candidates' of one combinator may be the result of running some nested combinator. Let me add a more explicit definition of that term to see if that helps, otherwise the text gets pretty unweildy.